pkg/machineproviders/providers/openshift/machine/v1beta1/failuredomain/failuredomain.go

```diff
@@ -90,16 +155,197 @@ func NewFailureDomains(failureDomains machinev1.FailureDomains) ([]FailureDomain
+func newGCPFailureDomains(failureDomains machinev1.FailureDomains) ([]FailureDomain, error) {
+	foundFailureDomains := []FailureDomain{}
+	if failureDomains.GCP == nil {
+		return foundFailureDomains, errMissingFailureDomain
+	}
+
+	for _, failureDomain := range *failureDomains.GCP {
+		foundFailureDomains = append(foundFailureDomains, NewGCPFailureDomain(failureDomain))
+	}
+
+	return foundFailureDomains, nil
+}
+
+// newOpenStackFailureDomains constructs a slice of OpenStack FailureDomain from machinev1.FailureDomains.
+func newOpenStackFailureDomains(failureDomains machinev1.FailureDomains) ([]FailureDomain, error) {
+	foundFailureDomains := []FailureDomain{}
+	if failureDomains.OpenStack == nil {
+		return foundFailureDomains, errMissingFailureDomain
+	}
+
+	for _, failureDomain := range *failureDomains.OpenStack {
+		foundFailureDomains = append(foundFailureDomains, NewOpenStackFailureDomain(failureDomain))
+	}
+
+	return foundFailureDomains, nil
+}
+
+// NewFailureDomainsFromMachines creates a slice of FailureDomains representing the failure domains of the provided machines.
+func NewFailureDomainsFromMachines(machines []*machinev1beta1.Machine, platform configv1.PlatformType) ([]FailureDomain, error) {
+	failureDomains := []FailureDomain{}
+
+	for _, machine := range machines {
+		failureDomain, err := newFailureDomainFromProviderSpec(machine.Spec.ProviderSpec.Value, platform)
+		if err != nil {
+			return nil, fmt.Errorf("error getting failure domain from machine %s: %w", machine.Name, err)
+		}
+
+		failureDomains = append(failureDomains, failureDomain)
+	}
+
+	return failureDomains, nil
+}
+
+// newFailureDomainFromProviderSpec creates a FailureDomain from machine providerSpec.
+func newFailureDomainFromProviderSpec(rawProviderSpec *runtime.RawExtension, platform configv1.PlatformType) (FailureDomain, error) {
+	var (
+		fd  FailureDomain
+		err error
+	)
+
+	if rawProviderSpec == nil {
+		return nil, errMachineMissingProviderSpec
+	}
+
+	switch platform {
+	case configv1.AWSPlatformType:
+		fd, err = newFailureDomainFromProviderSpecAWS(rawProviderSpec)
+	case configv1.AzurePlatformType:
+		fd, err = newFailureDomainFromProviderSpecAzure(rawProviderSpec)
+	case configv1.GCPPlatformType:
+		fd, err = newFailureDomainFromProviderSpecGCP(rawProviderSpec)
+	case configv1.OpenStackPlatformType:
+		fd, err = newFailureDomainFromProviderSpecOpenStack(rawProviderSpec)
+	default:
+		return nil, fmt.Errorf("%w: %s", errUnsupportedPlatformType, platform)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting failure domain from provider spec: %w", err)
+	}
+
+	return fd, nil
+}
+
+// newFailureDomainFromProviderSpecAWS creates a FailureDomain from AWS providerSpec.
+func newFailureDomainFromProviderSpecAWS(rawProviderSpec *runtime.RawExtension) (FailureDomain, error) {
+	providerSpec := machinev1beta1.AWSMachineProviderConfig{}
+	err := json.Unmarshal(rawProviderSpec.Raw, &providerSpec)
+
+	if err != nil {
```

Comment:
Nit, inline the error
```suggestion
	providerSpec := machinev1beta1.AWSMachineProviderConfig{}
	
	if err := json.Unmarshal(rawProviderSpec.Raw, &providerSpec); err != nil {
```